perf/reader.go

```diff
@@ -307,11 +316,16 @@ func (pr *Reader) SetDeadline(t time.Time) {
 // Returns os.ErrDeadlineExceeded if a deadline was set.
 func (pr *Reader) Read() (Record, error) {
 	var r Record
+
 	return r, pr.ReadInto(&r)
 }
 
 // ReadInto is like Read except that it allows reusing Record and associated buffers.
 func (pr *Reader) ReadInto(rec *Record) error {
+	if pr.overWritable && !pr.paused {
```

Comment:
> But it seems indeed better to have the whole function running under lock.

I think that's not possible since then calling .Pause would also block. Not sure how to fix that yet.